File: src/com/medium/Subsets_78.java
/**
 * 
 */
package com.medium;

import java.util.ArrayList;
import java.util.List;

/**
 * @author Chris Lin
 * @version 1.0
 */
public class Subsets_78 {
    // 2019年8月11日
    // Use backtrack
    // Time Complexity: O(n*2^n)
    // Space Complexity:O(n)
    // Runtime: 1 ms, faster than 44.44%
    // Memory Usage: 37.2 MB, less than 99.18%
    private List<List<Integer>> mResult = new ArrayList<List<Integer>>();

    public List<List<Integer>> subsets(int[] nums) {
        backTracking(nums, new ArrayList<>(), 0);

        return mResult;
    }

    private void backTracking(int[] nums, List<Integer> mTmp, int mStart) {
        mResult.add(new ArrayList<>(mTmp));

        for (int i = mStart; i < nums.length; i++) {
            if (!mTmp.contains(nums[i])) {
                mTmp.add(nums[i]);
                backTracking(nums, mTmp, i + 1);
                mTmp.remove(mTmp.size() - 1);
            }
        }
    }

    // 2019年8月11日
    // Use bit operation
    // Time Complexity: O(n*2^n)
    // Space Complexity:O(n)
    // Runtime: 1 ms, faster than 44.44%
    // Memory Usage: 37.2 MB, less than 99.18%
    public List<List<Integer>> subsets2(int[] nums) {
        for (int i = 0; i < Math.pow(2, nums.length); i++) {

            List<Integer> mTmp = new ArrayList<>();
            for (int j = 0; j < nums.length; j++) {
                if (((1 << j) & i) != 0) { // 0:not take, 1:take
                    mTmp.add(nums[j]);
                }
            }
            mResult.add(mTmp);
        }

        return mResult;
    }
}

File: Golang_src/medium/subsets_78.go
package medium

import "math"

// Use backtracking:
// Time Complexity: O(n*2^n)
// Space Complexity:O(n)
// Runtime: 2 ms, faster than 27.30%
// Memory Usage: 2.6 MB, less than 20.73%
func subsets(nums []int) [][]int {
	result := [][]int{}
	keep := []int{}
	size := len(nums)

	var dfs func(idx, subSize int)
	dfs = func(idx, subSize int) {
		if subSize == 0 {
			tmp := make([]int, len(keep))
			copy(tmp, keep)
			result = append(result, tmp)
			return
		}

		for i := idx; i < size; i++ {
			keep = append(keep, nums[i])
			dfs(i+1, subSize-1)
			keep = keep[:len(keep)-1]
		}
	}

	for i := 0; i <= size; i++ {
		dfs(0, i)
	}

	return result
}

// Use bitmask:
// Time Complexity: O(n*2^n)
// Space Complexity:O(n)
// Runtime: 0 ms, faster than 100.00%
// Memory Usage: 2.4 MB, less than 16.25%
func subsets2(nums []int) [][]int {
	size := len(nums)
	result := [][]int{}
	if size == 0 {
		return result
	}

	last := int(math.Pow(2.0, float64(size)))

	for i := 0; i < last; i++ {
		tmp := []int{}
		for j := 0; j < size; j++ {
			if (i & (1 << j)) != 0 {
				tmp = append(tmp, nums[j])
			}
		}
		result = append(result, tmp)
	}

	return result
}
